loader/compat_php.c

```diff
@@ -0,0 +1,212 @@
+#include "compat_php.h"
+#include "php_dd_library_loader.h"
+
+#include <php.h>
+#include <stdbool.h>
+
+#define HT_FLAGS(ht) (ht)->u.flags
+#define PHP_70_71_72_IS_STR_PERSISTENT (1 << 0)
+#define PHP_70_71_72_IS_STR_INTERNED (1 << 1)
+
+typedef struct _php70_71_72_zend_refcounted_h {
+	uint32_t         refcount;			/* reference counter 32-bit */
+	union {
+		struct {
+			ZEND_ENDIAN_LOHI_3(
+				zend_uchar    type,
+				zend_uchar    flags,    /* used for strings & objects */
+				uint16_t      gc_info)  /* keeps GC root number (or 0) and color */
+		} v;
+		uint32_t type_info;
+	} u;
+} php70_71_72_zend_refcounted_h;
+
+ZEND_API zval *ZEND_FASTCALL zend_hash_set_bucket_key(HashTable *ht, Bucket *b, zend_string *key) __attribute__((weak));
+
+ZEND_API zval *ZEND_FASTCALL zend_hash_update(HashTable *ht, zend_string *key, zval *pData) __attribute__((weak));
+ZEND_API zval *ZEND_FASTCALL _zend_hash_update(HashTable *ht, zend_string *key, zval *pData) __attribute__((weak));
+
+static bool ddloader_zstr_is_interned(int php_api_no, zend_string *key) {
+    if (php_api_no <= 20170718) {  // PHP 7.0 - 7.2
+        php70_71_72_zend_refcounted_h *gc = (php70_71_72_zend_refcounted_h*)&key->gc;
+        return (gc->u.v.flags & PHP_70_71_72_IS_STR_INTERNED);
```

Comment:
```suggestion
        return GC_TYPE_INFO(gc) & (PHP_70_71_72_IS_STR_INTERNED << 8);
```
Is probably the simpler way to express what you want. At least you're also 8-shifting for IS_STR_PERSISTENT below. Let's not have multiple ways to express the same thing.